feat: upgrade mode

jstuczyn opened the pull request:
## Ticket

[NET-341](https://nymtech.atlassian.net/browse/NET-341): [NET-452](https://nymtech.atlassian.net/browse/NET-452)

## Description

This PR introduces required changes to support the upgrade mode within the vpn client. The relevant changes include communication between the Account Controller and Bandwidth Controller to convey changes to the system state as well as introduction of the expllicit `UpgradeModeState` on the Account Controller.

for more detailed description please refer to the main PR on the monorepo: https://github.com/nymtech/nym/pull/6174 and in particular the attached confluence document: https://nymtech.atlassian.net/wiki/spaces/KB/pages/263880725/Implementation#

this PR depends on: https://github.com/nymtech/nym/pull/6174 https://github.com/nymtech/nym/pull/6189 and https://github.com/nymtech/websites/pull/653 to be merged before full compatibility is actually possible. 


## Checklist:

- [ ] Changelog

## Screenshots (optional, if UI related)


[NET-341]: https://nymtech.atlassian.net/browse/NET-341?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[NET-452]: https://nymtech.atlassian.net/browse/NET-452?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nymtech/nym-vpn-client/3932)
<!-- Reviewable:end -->

jstuczyn committed:
squashed feature/upgrade-mode (#3932)

introduced explicit details field to storage error

to give it clear separation from the context information

additional explicit error variant

moved definition of 'AvailableBandwidth'

rename client instance

missing implementation of 'remove_upgrade_mode_jwts'

rebased monorepo - updated refs + additional resultant linter fixes

adjust protobuf AccountControllerState state definition after rebasing

clippy

update monorepo dependency to use compatible published branch

post rebasing fixes

finished UpgradeModeState handlers

communication between AC and BC during UM

retrieving upgrade mode information when querying/topping-up bandwidth

transition from RequestingZkNymsState into UpgradeModeState

updated NymVpnZkNym response to include upgrade more related data

formatting

helpers to make checks more readable

DEV START: use local nym deps

simonwicky commented on nym-vpn-core/crates/nym-vpn-proto/proto/nym_vpn_service.proto:
```diff
@@ -645,6 +646,7 @@ message AccountControllerState {
     Error error = 4;
     Offline offline = 5;
     Decentralised decentralised = 6;
+    UpgradeMode upgrade_mode = 7;
```
`RequestingZkNyms requesting_zk_nyms = 7;` was added in `develop` and it might not show up when rebasing. Don't forget to update the number here then

simonwicky replied:
Outdated comment

jstuczyn replied:
already done it - CI was complaining so had no choice but to do it : D

simonwicky commented on nym-vpn-core/crates/nym-vpn-lib-types/src/account/request_zknym.rs:
```diff
@@ -201,17 +217,23 @@ impl From<RequestZkNymError> for RequestZkNymErrorReason {
             | RequestZkNymError::MissingPendingRequest(_)
-            | RequestZkNymError::Internal(_) => Self::Internal(err.to_string()),
+            | RequestZkNymError::Internal(_) | RequestZkNymError::MalformedUpgradeModeJWT
+            // TODO: not entirely sure if revocation fits here the best
```
I'd say it doesn't fit here, but it doesn't fit anywhere else. Maybe a new `RequestZkNymErrorReason` ?

simonwicky commented on nym-vpn-core/crates/nym-vpn-lib-types/src/account/controller_error.rs:
```diff
@@ -50,6 +51,16 @@ impl AccountControllerErrorStateReason {
         matches!(self, Self::ApiFailure { .. } | Self::Internal { .. })
     }
+
+    pub fn storage(context: impl Into<String>) -> AccountControllerErrorStateReason {
```
The idea behind the `context` part is to signal from where it was triggered, and `details` to explain what happened. 
While I have no problem with this, adding a `details` field to keep the `context` meaning consistent would be great. 
e.g. here : https://github.com/nymtech/nym-vpn-client/blob/0fd3c2c65953e3976d16f477da85f784dc6e82d3/nym-vpn-core/crates/nym-vpn-account-controller/src/state_machine/syncing_state/requesting_zknym_state.rs#L309

simonwicky commented on nym-vpn-core/crates/nym-vpn-lib/src/types.rs:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct AvailableBandwidth {
```
This `type.rs` seems a bit out of place. 
If it's only used by the BC, maybe define it there?
If it's gonna be used in multiple places, then `nym-vpn-lib-types` is the place

simonwicky replied:
Granted, `bandwidth_controller.rs` is a bit big, maybe a `bandwidth_controller/mod.rs`?

simonwicky commented on nym-vpn-core/crates/nym-vpn-lib/src/bandwidth_controller.rs:
```diff
@@ -389,77 +532,186 @@ impl BandwidthController {
+            .ticket_provider
+            .get_upgrade_mode_token()
+            .await
+            .map_err(|source| SpecificGatewayError::UpgradeModeTokenRequest {
```
If I'm reading this right, it's more of a storage error than a `SpecificGatewayError` one, because it would happen regardless of the gateway used, unlike a top up or a bandwidth query right?

simonwicky replied:
Ping @neacsu as well

neacsu replied:
Yes, I agree, this error layer is supposed to be used for differentiating between entry and exit specific errors

neacsu replied:
Also applies to `UnavailableUpgradeModeToken`  I would say

jstuczyn replied:
will change.

simonwicky commented on nym-vpn-core/crates/nym-vpn-lib/src/bandwidth_controller.rs:
```diff
@@ -389,77 +532,186 @@ impl BandwidthController {
+    }
+
+    async fn check_bandwidth(&mut self, entry: bool, current_period: Duration) -> Option<Duration> {
+        let wg_metadata_client = if entry {
```
nit: I know it was already the case, but that naming is wrong imo, could we call it `bw_client` instead? It's not always the metadata endpoint

simonwicky commented on nym-vpn-core/crates/nym-vpn-api-client/src/response.rs:
```diff
@@ -22,7 +23,7 @@ pub struct NymVpnRegisterAccountResponse {
     pub account_addr: String,
     pub status: NymVpnRegisterAccountStatusResponse,
-    pub account_token: String,
+    pub account_token: Option<String>,
```
Wait why is this an option now? What does that have to do with UM? Rebasing artefact?

jstuczyn replied:
because as I was running vpn api locally, it turns out, it doesn't always return the account_token

simonwicky replied:
ping @neacsu as I vaguely remember having the same issue at some point. Are we hitting the wrong endpoint again?

neacsu replied:
Might have been the case at some point, but now it's a `String` in the website repo, so this shouldn't be modified. (https://github.com/nymtech/websites/blob/1ec9396877edd465266e6d23992fa32f187b5ccb/packages/vpn-api-common/public-api-client/schemas.gen.ts#L3058)
<!-- Reviewable comment -Oe121q_78Ye0Mbb6tgs:bvvz8mp -->
<!-- Sent from Reviewable.io -->

simonwicky commented on nym-vpn-core/crates/nym-vpn-api-client/src/response.rs:
```diff
@@ -64,6 +65,28 @@ pub struct NymVpnAccountSummaryWithDeviceResponse {
     pub active_device: Option<NymVpnDevice>,
 }
 
+impl NymVpnAccountSummaryWithDeviceResponse {
```
Nice one :)

simonwicky commented on nym-vpn-core/crates/nym-vpn-account-controller/src/storage/credentials/mod.rs:
```diff
@@ -128,6 +131,39 @@ impl VpnCredentialStorage {
+    }
+
+    pub(crate) async fn remove_upgrade_mode_jwts(&self) -> Result<(), Error> {
+        todo!()
```
Is that `todo!()` supposed to be there?

jstuczyn replied:
absolutely not. i must have missed it. I'd have expected clippy to have caught that, but it seems that's not the case

simonwicky commented on nym-vpn-core/crates/nym-vpn-account-controller/src/commands/zknym_handler/request.rs:
```diff
@@ -98,24 +167,42 @@ impl RequestZkNymTask {
+            }
+            // unfortunately no error message is included, so we just replicate old behaviour
+            NymVpnZkNymStatus::Error => {
+                return Err(RequestZkNymError::MissingBlindedShares);
```
While that whole "checking the zk-nym status" part was very much missing and leading to weird errors before, we should definitely not return that error there. 
It should be a new variant, akin to the `Revoked` one above.

jstuczyn committed:
another set of post-rebasing fixes

jstuczyn committed:
updated raw data in NymVpnZkNym response parsing unit tests

the response has changed after the content of the upgrade mode attestation had been adjusted

neacsu reviewed (changes requested):
Can this be tracked by a VPN ticket as well, so that we make sure it goes through the VPN QA process as well?

> @neacsu reviewed 19 of 40 files at r1, 11 of 21 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3932#-:-Oe10gFn5o1K9ptUz2YC:bm1dhhp)* status: 30 of 45 files reviewed, 11 unresolved discussions (waiting on @doums, @jstuczyn, @mrkiwi, @pronebird, @rokas-ambrazevicius, and @simonwicky)

___
*[`nym-vpn-core/Cargo.toml` line 204 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3932#-Oe0wraH1Tj2mlCPvUhP:-Oe0wraH1Tj2mlCPvUhQ:b-c1oxbp) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/f97020dd2e7b048460e0dcb9df12f2cb311c9dd0/nym-vpn-core/Cargo.toml#L204)):*
> ```TOML
> 
> # For normal development
> nym-authenticator-client = { git = "https://github.com/nymtech/nym", branch = "feature/vpn-upgrade-mode-compat" }
> ```

Can we get this to be some release branch again?


<!-- Sent from Reviewable.io -->

neacsu reviewed (changes requested):
> @neacsu reviewed 5 of 40 files at r1, 10 of 21 files at r2.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3932)* status: all files reviewed, 12 unresolved discussions (waiting on @doums, @jstuczyn, @mrkiwi, @pronebird, @rokas-ambrazevicius, and @simonwicky)

___
*[`nym-vpn-core/crates/nym-wg-metadata-client/src/types.rs` line 7 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3932#-Oe16KbE52oImyhFvm4z:-Oe16KbE52oImyhFvm5-:bs58921) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/f97020dd2e7b048460e0dcb9df12f2cb311c9dd0/nym-vpn-core/crates/nym-wg-metadata-client/src/types.rs#L7)):*
> ```XML
> 
> #[derive(Debug, Clone, Copy)]
> pub struct AvailableBandwidth {
> ```

I think all this file should be part of the `nym_wireguard_private_metadata_client` crate and hide all the version implementation details


<!-- Sent from Reviewable.io -->

jstuczyn committed:
removed option wrapper around 'account_token' since it can't be triggered in realistic scenarios

jstuczyn committed:
moved 'AvailableBandwidth' definition to the monorepo

jstuczyn committed:
updated error variants for upgrade mode rechecks

neacsu reviewed (changes requested):
> @neacsu reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3932)* status: all files reviewed, 12 unresolved discussions (waiting on @doums, @jstuczyn, @mrkiwi, @neacsu, @pronebird, @rokas-ambrazevicius, and @simonwicky)

___
*[`nym-vpn-core/crates/nym-wg-metadata-client/src/lib.rs` line 20 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3932#-Oe1VyJ87gvVhsgYteUk:-Oe1VyJ92Cr4PYaM7oPA:b9stf09) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/f97020dd2e7b048460e0dcb9df12f2cb311c9dd0/nym-vpn-core/crates/nym-wg-metadata-client/src/lib.rs#L20)):*
> ```XML
> 
> pub mod error;
> pub mod types;
> ```

You should also remove the `types.rs` file


<!-- Sent from Reviewable.io -->

jstuczyn committed:
purge AvailableBandwidth for real this time

jstuczyn committed:
chore: update monorepo ref as corresponding PR got merged
